src/Session.cpp

```diff
@@ -0,0 +1,1456 @@
+/*
+ *  Copyright (C) 2022 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "Session.h"
+
+#include "aes_decrypter.h"
+#include "parser/DASHTree.h"
+#include "parser/HLSTree.h"
+#include "parser/SmoothTree.h"
+#include "samplereader/ADTSSampleReader.h"
+#include "samplereader/DummySampleReader.h"
+#include "samplereader/FragmentedSampleReader.h"
+#include "samplereader/SubtitleSampleReader.h"
+#include "samplereader/TSSampleReader.h"
+#include "samplereader/WebmSampleReader.h"
+#include "utils/Base64Utils.h"
+#include "utils/SettingsUtils.h"
+#include "utils/StringUtils.h"
+#include "utils/Utils.h"
+#include "utils/log.h"
+
+#include <kodi/addon-instance/Inputstream.h>
+
+using namespace UTILS;
+using namespace kodi::tools;
+
+static const AP4_Track::Type TIDC[adaptive::AdaptiveTree::STREAM_TYPE_COUNT] = {
+    AP4_Track::TYPE_UNKNOWN, AP4_Track::TYPE_VIDEO, AP4_Track::TYPE_AUDIO,
+    AP4_Track::TYPE_SUBTITLES};
+
+
+CSession::CSession(const PROPERTIES::KodiProperties& kodiProps,
+                   const std::string& manifestUrl,
+                   const std::map<std::string, std::string>& mediaHeaders,
+                   const std::string& profilePath)
+  : m_kodiProps(kodiProps),
+    m_manifestUrl(manifestUrl),
+    m_mediaHeaders(mediaHeaders),
+    m_dummySampleReader(std::make_unique<CDummySampleReader>()),
+    m_KodiHost(std::make_unique<CKodiHost>()),
+    m_reprChooser(CHOOSER::CreateRepresentationChooser(kodiProps))
+
+{
+  m_KodiHost->SetProfilePath(profilePath);
+
+  switch (kodiProps.m_manifestType)
+  {
+    case PROPERTIES::ManifestType::MPD:
+      m_adaptiveTree =
+          std::make_unique<adaptive::DASHTree>(adaptive::DASHTree(kodiProps, m_reprChooser.get()));
```

Comment:
i think that you have done a bit of mess with unique ptr's
you are using unique ptr's and sharing his raw pointers i think this is a bad implementation

I would rather leave everything as it was originally with raw pointers
and with a later PR change them with a shared_ptr instead of unique ptr
which has exactly the purpose of being shared safely

Reply:
> i think that you have done a bit of mess with unique ptr's
Yeah sorry, I did make a mess. I've removed for chooser because yes it was being abused, but do you think it's appropriate to leave any of the adaptiveTree/dllHelper/KodiHost ones as unique_ptr?

Reply:
dllHelper and KodiHost unique_ptr are ok
but m_adaptiveTree not better keep raw for now and change to sharedptr in future prs